src/interactable_checkpoint.c

```diff
@@ -0,0 +1,276 @@
+    disp->focused = 0;
+    disp->unk28->unk0 = -1;
+    disp->unk10 = 0x2000;
+
+    zone = LEVEL_TO_ZONE(gCurrentLevel);
+    anim = sAnimIdsCheckpoint[zone][0];
+    variant = sAnimIdsCheckpoint[zone][1];
+
+    chkPt->task = sub_8009628(0x2000, anim, variant, Task_8063228);
+}
+
+void Task_8062FD8(void)
+{
+    Sprite_Checkpoint *chkPt = TaskGetStructPtr(gCurTask);
+    Sprite *disp = &chkPt->displayed;
+    Interactable *ia = chkPt->base.ia;
+    s32 posX, posY;
+    posX = SpriteGetScreenPos(chkPt->base.spriteX, chkPt->base.regionX);
+    posY = SpriteGetScreenPos(ia->y, chkPt->base.regionY);
+
+    disp->x = posX - gCamera.x;
+    disp->y = posY - gCamera.y;
+
+    if (IS_OUT_OF_CAM_RANGE(disp->x, disp->y)) {
+        ia->x = chkPt->base.spriteX;
+        TaskDestroy(chkPt->task);
+        TaskDestroy(gCurTask);
+    } else {
+        if (!(gPlayer.moveState & (MOVESTATE_400000 | MOVESTATE_DEAD))
+            && posX <= Q_24_8_TO_INT(gPlayer.x)) {
+            gPlayer.checkPointX = gUnknown_080D63FC[gCurrentLevel][0];
+            gPlayer.checkPointY = gUnknown_080D63FC[gCurrentLevel][1];
+            gPlayer.checkpointTime = gUnknown_030053E4;
+
+            if (gUnknown_030055B0 == 0)
+                gUnknown_030055B0++;
+
+            disp->graphics.anim = SA2_ANIM_CHECKPOINT;
+            disp->variant = SA2_ANIM_VARIANT_CHECKPOINT_HIT;
+            disp->unk21 = 0xFF;
+
+            gCurTask->main = Task_8063108;
+
+            m4aSongNumStart(SE_CHECKPOINT);
+        }
+
+        sub_8004558(disp);
+        sub_80051E8(disp);
+    }
+}
+
+void Task_8063108(void)
+{
+    Sprite_Checkpoint *chkPt = TaskGetStructPtr(gCurTask);
+    Sprite *disp = &chkPt->displayed;
+    Interactable *ia = chkPt->base.ia;
+    s32 posX, posY;
+    posX = SpriteGetScreenPos(chkPt->base.spriteX, chkPt->base.regionX);
+    posY = SpriteGetScreenPos(ia->y, chkPt->base.regionY);
+
+    disp->x = posX - gCamera.x;
+    disp->y = posY - gCamera.y;
+
+    if (IS_OUT_OF_CAM_RANGE(disp->x, disp->y)) {
+        ia->x = chkPt->base.spriteX;
+        TaskDestroy(chkPt->task);
+        TaskDestroy(gCurTask);
+    } else {
+        if (sub_8004558(disp) == 0) {
+            gCurTask->main = Task_806319C;
+        }
+
+        sub_80051E8(disp);
+    }
+}
+
+void Task_806319C(void)
+{
+    Sprite_Checkpoint *chkPt = TaskGetStructPtr(gCurTask);
+    Sprite *disp = &chkPt->displayed;
+    Interactable *ia = chkPt->base.ia;
+    s32 posX, posY;
+    posX = SpriteGetScreenPos(chkPt->base.spriteX, chkPt->base.regionX);
+    posY = SpriteGetScreenPos(ia->y, chkPt->base.regionY);
+
+    disp->x = posX - gCamera.x;
+    disp->y = posY - gCamera.y;
+
+    if (IS_OUT_OF_CAM_RANGE(disp->x, disp->y)) {
+        ia->x = chkPt->base.spriteX;
+        TaskDestroy(chkPt->task);
+        TaskDestroy(gCurTask);
+    }
+}
+
+// static
+void TaskDestructor_8063214(struct Task *t)
+{
+    Sprite_Checkpoint *chkPt = TaskGetStructPtr(t);
+    void *gfx = chkPt->displayed.graphics.dest;
```

Comment:
This has to be set for this functiont to be called right? What's the reason to assume that this TaskDestructor could be called within this being initialised?

Reply:
Writing it separately is not just more readable, but also more debuggable.

I guess an assert would be better, but... generally you should check whether a pointer is NULL before calling any "free" function, right?